Add July knowledge engineering posts (#10)
input-output-hk/engineering: blog/2022-07-26-ghcjs-linker.md

## Introduction

I recently gave a short presentation on the workings of the GHCJS linker. This post is a summary of the content.

## JavaScript "executables"

The task of a linker is collecting and organizing object files and resources into a loadable library or executable program. JavaScript can be run in various environments, for example the browser or node.js, and not in all of these the concept of an executable makes sense.

Therefore, when we link a Haskell program, we generate a `jsexe` directory filled with various files that allow us to run the JavaScript result:

| File        | Description          |
| :----:        | :---:             |
| `out.js`      | compiled/linked Haskell code          |
| `out.frefs.*` | list of foreing calls from `out.js` |
| `out.stats`   | source code size origin statistics for `out.js` |
| `lib.js`      | non-Haskell code, from `js-sources` in packages and RTS. possibly preprocessed |
| `rts.js`      | generated part of RTS (apply functions and similarly repetitive things) |
| `runmain.js`  | single line just starts `main` |
| `all.js`      | complete runnable program, created by combining `out.js`, `lib.js`, `rts.js` and `runmain.js` |

> Most of the work done by the linker is producing `out.js`, and that's what we'll be focussing on in the next sections.

Reviewer: `focussing` is a misspelled version of `focusing`

From a high level, the procedure that the linker follow is this:

| Step |
| :---: |
| Read object files from dependencies into memory |
| Decode dependency part of all object files in dependencies (includes reading the string tables) |
| Using dependency data, find all code reachable from `main` |
| Decode reachable binding groups |
| Render AST to JavaScript |
| Construct initializers from metadata | 

We avoid decoding (deserializing) the binding groups that do end up in the linked result to keep the memory consumption lower. Still the linker requires a lot of memory for larger programs, so we may need to make more improvements in the future.

## The Compactor

The compactor is an optional link-time transformation step that reduces code size. It consists of a lightweight (i.e. no expensive operations like dataflow analysis) rewrite of the code contained in the object files. The compactor is disabled when linking with the `-debug` flag. There are a few steps involved.

### Renaming private symbols

> Haskell modules have a bunch of public names, i.e. those exported from the module, and private names, non-exported bindings and new bindings introduced by the optimizer or other parts of the compilation pipeline.

Reviewer 2: I would stress the motivation for the transformation more: could you add something like

```suggestion
Haskell names are quite long by default: they need to be globally unique, hence they contain their defining unit-id and module name. For example: `mtl-2.2.2-somehash-Control.Monad.State.Lazy.execState_go1` (special characters would be z-encoded but it isn't shown here).
At link time, however, the compactor can replace all these names with shorter ones as long as it maintains their global uniqueness.

Haskell modules have a bunch of public names, i.e. those exported from the module, and private names, i.e. non-exported bindings and new bindings introduced by the optimizer or other parts of the compilation pipeline.
The compactor must only rename private symbols to still allow incremental linking (see below) and to avoid renaming "wired-in" symbols that are explicitly used in hand-written JavaScript code (e.g. in the RTS) that isn't processed by the compactor.
```

Side note: as we have full control over the linking process, perhaps we could also rename exposed symbols as long as we keep a table associating old names to new ones in the bundle for incremental linking and if we used placeholder for wired-in names in the JavaScript files (e.g. CPP constants as we do CPP at link time too)?

Private symbols are only referred to from within the same module. It doesn't matter witch JavaScript name we pick for them, as long as there are no overlaps between the names from different modules. The compactor renames all the private symbols using a global sequence to ensure short names that do not overlap.

### Block Initializer

Without the compactor, the linker generates an `h$initObj` initialization call (or `h$o`) call for each global Haskell heap value. The code for this can get quite big. The compactor collects all heap objects to be initialized in a single large array and encodes the metadata in a string. This makes the initialization code much more compact.

> 

Reviewer 2: It would be nice to add a code example here showing both approaches

### Deduplication

An optional step in the compactor is deduplication of code. When deduplication is enabled with the `-dedupe` flag, the compactor looks for functionally equivalent pieces of JavaScript in the output and merges them. This can result in a significant reduction of code size.

## Incremental Linking

> The linker supports building programs that are loaded incrementally. This is used for example for Template Haskell. The process that runs the Template Haskell stays alive during compilation of a whole module. When the first Template Haskell expression is compiled, it is linked against all its dependencies and the RTS and sent over to be run in the evaluator process.

Reviewer 2: ```suggestion
The linker supports building programs that are loaded incrementally. This is used for example for Template Haskell. The process that runs the Template Haskell stays alive during compilation of a whole module. When the first Template Haskell expression is compiled, it is linked against all its dependencies (including the RTS) and the resulting JavaScript code is sent over to be run in the evaluator process.
```

> Subsequent Template Haskell expressions are evaluated in the same process, where the RTS is already loaded by then, and also the dependencies of the earlier Template Haskell expressions. Therefore each expression is only linked against dependecies that are not already loaded in the evaluator process. The linker keeps track of which dependencies have already been linked.

Reviewer 2: ```suggestion
As subsequent Template Haskell expressions are evaluated in the same process, there is no need to load already loaded dependencies (including the RTS) again and it is much more efficient to avoid doing so. Therefore the linker keeps track of which dependencies have already been linked and each subsequent TH expression is only linked against dependencies that are not already loaded in the evaluator process.
```
